scipy/special/_logsumexp.py

```diff
@@ -212,4 +212,65 @@ def softmax(x, axis=None):
     """
 
     # compute in log space for numerical stability
-    return np.exp(x - logsumexp(x, axis=axis, keepdims=True))
+    # subtraction can cause cancellation when x and logsumexp(x) are close to each other.
+    # return np.exp(x - logsumexp(x, axis=axis, keepdims=True))
+    return np.exp(log_softmax(x, axis=axis))
+
+
+def log_softmax(x, axis=None):
+    r"""
+    Logarithm of softmax function::
+
+        log_softmax(x) = log(softmax(x))
+
+    Parameters
+    ----------
+    x : array_like
+        Input array.
+    axis : int or tuple of ints, optional
+        Axis to compute values along. Default is None and softmax will be
+        computed over the entire array `x`.
+
+    Returns
+    -------
+    s : ndarray
```

Comment:
Does it return a scalar if the input is a scalar?

Reply:
Yes. I changed the docstring accordingly.